source/common/http/http1/codec_impl.cc

```diff
@@ -304,7 +304,17 @@ void StreamEncoderImpl::resetStream(StreamResetReason reason) {
   connection_.onResetStreamBase(reason);
 }
 
-void StreamEncoderImpl::readDisable(bool disable) { connection_.readDisable(disable); }
+void StreamEncoderImpl::readDisable(bool disable) {
+  if (disable) {
+    ++read_disable_calls_;
```

Comment:
Is there an expected max value for read_disable_calls_ that we can ASSERT?  At first glance, I think the answer is 1, the call from ServerConnectionImpl::onMessageComplete

Reply:
I don't think there's a fixed limit but it should be small.  Offhand I think the worst situation would be 
1) upstream http/2 window getting backed up
2) upstream CONNECTION getting backed up
3) hedging resulting in doubling the above.

Reply:
But those operate on the connection level readDisable counter.  It seems like this new method is only called from ServerConnectionImpl::onMessageComplete

Reply:
This function already existed.  Generally stream disabling is done by the peer.  For server streams if the router gets overrun, it tells the HCM, which readDisables the downstream stream to stop data from the client, and when the router gets drained the reverse happens to resume reading.  Ditto if the client is draining slowly - in this case it goes from the HCM through router's HttpUpstream which calls readDisable on the upstream stream

The special server-only logic in the codec is to disable reading between (maybe) pipelined requests.

Reply:
You're right, I was confused.